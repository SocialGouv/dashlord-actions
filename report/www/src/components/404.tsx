import * as React from "react";

import Table from "@codegouvfr/react-dsfr/Table";

import { Panel } from "./Panel";

type Report404Props = { data: Wget404Report };

export const Report404: React.FC<Report404Props> = ({ data }) => {
  const tableData = data && [
    ["destination"],
    ...data.map((url2) => [url2.link]),
  ];
  return (
    (data && data.length && (
      <Panel title="Erreurs 404" info="Pages introuvables">
        <Table data={tableData} />
      </Panel>
    )) ||
    null
  );
};
