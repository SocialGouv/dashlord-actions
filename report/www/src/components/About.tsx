import * as React from "react";

import CallOut from "@codegouvfr/react-dsfr/CallOut";

export const About: React.FC = () => (
  <CallOut className="fr-mb-3w" title="À propos">
    DashLord est <strong>100% open source</strong>, vous pouvez y contribuer en
    créant de la documentation, en proposant du code, ou encore en partageant
    vos questions{" "}
    <a
      href="https://github.com/socialgouv/dashlord"
      target="_blank"
      rel="noopener noreferrer"
    >
      sur le repository github
    </a>
    .
    <br /> <br />
    L&apos;objectif de DashLord est d&apos;identifier des points
    d&apos;attention mais aussi de proposer des solutions actionnables.
    <br /> <br />
    DashLord est né à la{" "}
    <a
      href="https://fabrique.social.gouv.fr"
      target="_blank"
      rel="noopener noreferrer"
    >
      {" "}
      Fabrique des ministères sociaux
    </a>{" "}
    pour répondre aux besoins d&apos;évaluation et de mise en oeuvre des bonnes
    pratiques de développement web.
    <br /> <br />
    Vous pouvez facilement créer votre propre DashLord pour vos URLs en suivant{" "}
    <a
      href="https://github.com/SocialGouv/dashlord/blob/main/README.md"
      target="_blank"
      rel="noopener noreferrer"
    >
      les instructions sur le GitHub
    </a>
    .
  </CallOut>
);
