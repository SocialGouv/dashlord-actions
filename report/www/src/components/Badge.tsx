import classNames from "classnames";
import React, { CSSProperties } from "react";
import { useRouter } from "next/router";

import styles from "./badge.module.scss";

type BadgeProps = {
  variant: string;
  className?: string;
  to?: string;
  style?: CSSProperties;
  external?: boolean;
  children: any;
};

const Badge: React.FC<BadgeProps> = ({
  children,
  external = false,
  variant,
  className,
  to,
  style,
}) => {
  const router = useRouter();
  return (
    <button
      type="button"
      className={classNames(className, styles[variant])}
      onClick={() => {
        if (to) {
          if (external) {
            window.open(to);
            return;
          }
          router.push(to);
        }
      }}
      style={style}
    >
      {children}
    </button>
  );
};

export default Badge;
