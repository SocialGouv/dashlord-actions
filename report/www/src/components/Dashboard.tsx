import * as React from "react";
import { Table } from "@dataesr/react-dsfr";
import { Search, Slash } from "react-feather";
import Link from "next/link";
import { format } from "date-fns";

import { AccessibilityWarnings } from "../lib/lighthouse/AccessibilityWarnings";
import {
  isToolEnabled,
  letterGradeValue,
  smallUrl,
  slugifyUrl,
} from "../utils";
import { Grade } from "./Grade";
import ColumnHeader from "./ColumnHeader";
import styles from "./dashboard.module.scss";
import { getLatestPhase } from "./BetagouvInfo";

type DashboardProps = { report: DashLordReport };

const IconUnknown = () => <Slash size={20} />;

const percent = (num: number | undefined): string =>
  (num !== undefined && `${Math.floor(num * 100)} %`) || "-";

const GradeBadge = ({
  grade,
  label,
  warning,
  to,
  colorVariant,
  style,
}: {
  grade: string | undefined;
  label?: string | number | undefined;
  warning?: string;
  to?: string;
  colorVariant?: ColorVariant;
  style?: React.CSSProperties;
}) => (
  <div style={{ textAlign: "center" }}>
    {grade || label ? (
      <Grade
        colorVariant={colorVariant}
        small
        warning={warning}
        grade={grade || `${label}`}
        label={label}
        to={to}
        style={style}
      />
    ) : (
      <IconUnknown />
    )}
  </div>
);

type GetColumnProps = {
  id: string;
  title: string;
  info: string;
  warning?: JSX.Element | undefined;
  hash: string;
  gradeKey: string;
  sort?: Function;
  category?: string;
  gradeLabel?: (s: UrlReport) => string | number | undefined;
  gradeStyle?: React.CSSProperties;
  colorVariant?: ColorVariant;
  getColorVariant?: (s: UrlReport) => ColorVariant;
  warningText?: (s: UrlReportSummary) => string | undefined;
};

export const Dashboard: React.FC<DashboardProps> = ({ report }) => {
  const getSummaryData = (rowData, grade) => {
    const { summary } = rowData as UrlReport;
    return (summary[grade] && letterGradeValue(summary[grade])) || -1;
  };

  //console.log(report);
  const sortSSLGrades = (a, b) =>
    b.summary.testsslExpireSoon - a.summary.testsslExpireSoon ||
    getSummaryData(a, "testsslGrade") - getSummaryData(b, "testsslGrade") ||
    1;

  const getColumn = ({
    id,
    title,
    info,
    sort,
    warning,
    hash,
    gradeKey,
    category,
    gradeLabel,
    warningText,
    colorVariant,
    getColorVariant,
    gradeStyle,
  }: GetColumnProps) => ({
    name: id,
    sortable: true,
    sort: (a, b) =>
      sort
        ? sort(a, b)
        : getSummaryData(a, gradeKey) - getSummaryData(b, gradeKey),
    label: title,
    headerRender: () => (
      <ColumnHeader title={title} info={info} warning={warning} />
    ),
    render: (rowData) => {
      const { summary } = rowData as UrlReport;
      return (
        <GradeBadge
          colorVariant={
            getColorVariant ? getColorVariant(rowData) : colorVariant
          }
          style={gradeStyle}
          grade={summary[gradeKey]}
          label={gradeLabel && gradeLabel(rowData)}
          warning={warningText && warningText(summary)}
          to={`/url/${encodeURIComponent(
            slugifyUrl((rowData as UrlReport).url)
          )}/${category ? `${category}/` : ""}#${hash}`}
        />
      );
    },
  });

  const lightHouseColumn = (id, title, info) =>
    getColumn({
      id,
      title,
      info,
      category: "best-practices",
      warning: id === "accessibility" ? <AccessibilityWarnings /> : undefined,
      hash: "lighthouse",
      gradeKey: `lighthouse_${id}Grade`,
      sort: (a, b) => {
        if (a.summary[`lighthouse_${id}`] === undefined) {
          return -1;
        }
        if (b.summary[`lighthouse_${id}`] === undefined) {
          return 1;
        }
        return (
          parseFloat(a.summary[`lighthouse_${id}`]) -
          parseFloat(b.summary[`lighthouse_${id}`])
        );
      },
      gradeLabel: (rowData) =>
        rowData.lhr && percent(rowData.summary[`lighthouse_${id}`]),
    });

  let columns = [
    {
      name: "url",
      label: `URL ${report && `(${report.length})`}`,
      sortable: true,
      render: (rowData) => (
        <div
          style={{
            width: "95%",
            overflow: "hidden",
            whiteSpace: "nowrap",
            textOverflow: "ellipsis",
          }}
        >
          <Link
            prefetch={false}
            href={`/url/${encodeURIComponent(
              slugifyUrl((rowData as UrlReport).url)
            )}`}
          >
            <a>
              <Search size={16} />
              &nbsp;
              {smallUrl((rowData as UrlReport).url)}
            </a>
          </Link>
        </div>
      ),
    },
  ];

  if (isToolEnabled("betagouv")) {
    columns.push(
      getColumn({
        id: "se_current_phase",
        title: "Phase",
        info: "Phase actuelle de la Startup d'Etat",
        hash: "betagouv",
        category: "informations",
        gradeKey: "seCurrentPhase",
        gradeStyle: { width: 140 },
        colorVariant: "info",
        sort: (a, b) => {
          // ensure we use latest phase for sorting
          if (!a.betagouv) return 1;
          if (!b.betagouv) return -1;
          return (
            getLatestPhase(b.betagouv.attributes.phases).index -
            getLatestPhase(a.betagouv.attributes.phases).index
          );
        },
        gradeLabel: (rowData) =>
          rowData.betagouv &&
          getLatestPhase(rowData.betagouv.attributes.phases).label,
      })
    );
  }

  if (isToolEnabled("declaration-a11y")) {
    columns.push(
      getColumn({
        id: "declaration-a11y",
        title: "Déclaration d'accessibilité",
        info: "Présence de la mention de conformité et de la déclaration",
        hash: "declaration-a11y",
        gradeKey: "declaration-a11y",
        category: "best-practices",
      })
    );
  }

  if (isToolEnabled("github_repository")) {
    columns.push(
      getColumn({
        id: "github_repository",
        title: "Code ouvert",
        info: "Le code est-il ouvert ?",
        hash: "github_repository",
        category: "best-practices",
        gradeKey: "githubRepositoryGrade",
        gradeLabel: (rowData) => {
          //@ts-ignore
          const count = rowData.summary.githubRepositoryGrade;
          if (count === "A") {
            return "✔";
          }
          return count;
        },
      })
    );
  }

  if (isToolEnabled("declaration-rgpd")) {
    columns.push(
      getColumn({
        id: "declaration-rgpd-ml",
        title: "Mentions légales",
        info: "Présence des mentions légales et conformité de celles-ci",
        hash: "declaration-rgpd",
        gradeKey: "declaration-rgpd-ml",
        category: "best-practices",
        gradeLabel: (rowData) => {
          switch (rowData.summary["declaration-rgpd-ml"]) {
            case "A":
              return "✔";
            case "D":
              return "";
            case "F":
              return "";
          }
        },
        warningText: (summary) =>
          (summary["declaration-rgpd-ml"] === "D" &&
            "Les mentions légales sont présentes mais incomplètes. Consultez les détails pour plus d'informations") ||
          (summary["declaration-rgpd-ml"] === "F" &&
            "Les mentions légales n'ont pas été détectées"),
      })
    );
    columns.push(
      getColumn({
        id: "declaration-rgpd-pc",
        title: "Politique de confidentialité",
        info: "Présence de la politique de confidentialité et conformité de celle-ci",
        hash: "declaration-rgpd",
        gradeKey: "declaration-rgpd-pc",
        category: "best-practices",
        gradeLabel: (rowData) => {
          switch (rowData.summary["declaration-rgpd-pc"]) {
            case "A":
              return "✔";
            case "D":
              return "";
            case "F":
              return "";
          }
        },
        warningText: (summary) =>
          (summary["declaration-rgpd-pc"] === "D" &&
            "La politique de confidentialité est présente mais incomplète. Consultez les détails pour plus d'informations") ||
          (summary["declaration-rgpd-pc"] === "F" &&
            "La politique de confidentialité n'a pas été trouvée"),
      })
    );
  }

  if (isToolEnabled("ecoindex")) {
    columns.push(
      getColumn({
        id: "ecoindex",
        title: "Eco-index",
        info: "Note éco-index",
        hash: "ecoindex",
        category: "best-practices",
        gradeKey: "ecoindexGrade",
      })
    );
  }

  if (isToolEnabled("lighthouse")) {
    columns = columns.concat([
      getColumn({
        id: "accessibility",
        title: "Tests auto accessibilité",
        info: "Bonnes pratiques en matière d'accessibilité web (LightHouse)",
        category: "best-practices",
        warning: <AccessibilityWarnings />,
        hash: "lighthouse",
        gradeKey: `lighthouse_accessibilityGrade`,
        sort: (a, b) => {
          if (a.summary[`lighthouse_accessibility`] === undefined) {
            return -1;
          }
          if (b.summary[`lighthouse_accessibility`] === undefined) {
            return 1;
          }
          return (
            parseFloat(a.summary[`lighthouse_accessibility`]) -
            parseFloat(b.summary[`lighthouse_accessibility`])
          );
        },
        getColorVariant: (rowData) => {
          const value =
            rowData.lhr && rowData.summary[`lighthouse_accessibility`];
          if (value >= 1) {
            return "success";
          } else if (value >= 0.85) {
            return "warning";
          } else if (value >= 0.7) {
            return "danger";
          } else {
            return "danger";
          }
        },
        gradeLabel: (rowData) => {
          const value =
            rowData.lhr && rowData.summary[`lighthouse_accessibility`];
          if (value >= 1) {
            return "✔";
          } else if (value >= 0.85) {
            return "B";
          } else if (value >= 0.7) {
            return "C";
          } else {
            return "⚠️";
          }
        },
      }),
      lightHouseColumn(
        "performance",
        "Performance",
        "Performances de chargement des pages web (LightHouse)"
      ),
      lightHouseColumn(
        "seo",
        "SEO",
        "Bonnes pratiques en matière de référencement naturel (LightHouse)"
      ),
    ]);
  }

  if (isToolEnabled("testssl")) {
    columns.push(
      getColumn({
        id: "ssl",
        title: "SSL",
        info: "Niveau de confiance du certificat SSL (testssl.sh)",
        hash: "testssl",
        gradeKey: "testsslGrade",
        gradeLabel: (rowData) => rowData.summary.testsslGrade,
        sort: sortSSLGrades,
        category: "securite",
        warningText: (summary) =>
          (summary.testsslExpireSoon &&
            summary.testsslExpireDate &&
            `Expire le : ${format(
              new Date(summary.testsslExpireDate),
              "dd/MM/yyyy"
            )}`) ||
          undefined,
      })
    );
  }

  if (isToolEnabled("http")) {
    columns.push(
      getColumn({
        id: "http",
        title: "HTTP",
        info: "Bonnes pratiques de configuration HTTP (Mozilla observatory)",
        hash: "http",
        category: "securite",
        gradeKey: "httpGrade",
      })
    );
  }

  if (isToolEnabled("updownio")) {
    columns = columns.concat([
      getColumn({
        id: "updownio",
        title: "Disponibilité",
        info: "Disponibilité du service (updown.io)",
        hash: "updownio",
        gradeKey: "uptimeGrade",
        category: "disponibilite",
        sort: (a, b) => {
          if (a.summary[`uptime`] === undefined) {
            return -1;
          }
          if (b.summary[`uptime`] === undefined) {
            return 1;
          }
          return a.summary[`uptime`] - b.summary[`uptime`];
        },
        gradeLabel: (rowData) =>
          rowData.summary.uptime !== undefined &&
          percent((rowData.summary.uptime || 0) / 100),
      }),
      getColumn({
        id: "updownio2",
        title: "Apdex",
        info: "Apdex: Application Performance Index : indice de satisfaction des attentes de performance (updown.io)",
        hash: "updownio",
        gradeKey: "apdexGrade",
        category: "disponibilite",
        gradeLabel: (rowData) => rowData.summary.apdex,
      }),
    ]);
  }

  if (isToolEnabled("dependabot")) {
    columns.push(
      getColumn({
        id: "dependabot",
        title: "Vulnérabilités",
        info: "Vulnérabilités applicatives detectées dans les dépendances du code (dependabot)",
        hash: "dependabot",
        category: "securite",
        gradeKey: "dependabotGrade",
        sort: (a, b) =>
          parseFloat(a.summary.dependabotCount || 0) -
          parseFloat(b.summary.dependabotCount || 0),
        gradeLabel: (rowData) => rowData.summary.dependabotCount,
      })
    );
  }

  if (isToolEnabled("codescan")) {
    columns.push(
      getColumn({
        id: "codescan",
        title: "CodeScan",
        info: "Potentielles vulnérabilités ou erreurs detectées dans les codes sources (codescan)",
        category: "securite",
        hash: "codescan",
        gradeKey: "codescanGrade",
        sort: (a, b) =>
          parseFloat(a.summary.codescanCount || 0) -
          parseFloat(b.summary.codescanCount || 0),
        gradeLabel: (rowData) => {
          const count = rowData.summary.codescanCount;
          if (count === 0) {
            return "✔";
          }
          return count;
        },
      })
    );
  }

  if (isToolEnabled("nmap")) {
    columns = columns.concat([
      getColumn({
        id: "nmap",
        title: "Nmap",
        info: "Vulnérabilités réseau detectées par Nmap",
        category: "securite",
        hash: "nmap",
        gradeKey: "nmapGrade",
      }),
      getColumn({
        id: "nmap2",
        title: "Ports ouverts",
        info: "Ports TCP ouverts détectés par nmap",
        category: "securite",
        sort: (a, b) => {
          if (a.summary[`nmapOpenPortsCount`] === undefined) {
            return -1;
          }
          if (b.summary[`nmapOpenPortsCount`] === undefined) {
            return 1;
          }
          return (
            a.summary[`nmapOpenPortsCount`] - b.summary[`nmapOpenPortsCount`]
          );
        },
        hash: "nmap",
        gradeKey: "nmapOpenPortsGrade",
        gradeLabel: (rowData) => rowData.summary.nmapOpenPortsCount,
      }),
    ]);
  }

  if (isToolEnabled("thirdparties")) {
    columns = columns.concat([
      getColumn({
        id: "trackers",
        title: "Trackers",
        info: "Nombre de scripts externes détectés",
        warning: (
          <div>
            Certains scripts externes légitimes peuvent être considérés comme
            trackers.
          </div>
        ),
        category: "best-practices",
        hash: "thirdparties",
        gradeKey: "trackersGrade",
        sort: (a, b) => {
          if (a.summary[`trackersCount`] === undefined) {
            return -1;
          }
          if (b.summary[`trackersCount`] === undefined) {
            return 1;
          }
          return a.summary[`trackersCount`] - b.summary[`trackersCount`];
        },
        gradeLabel: (rowData) => {
          const count = rowData.summary.trackersCount;
          if (count === 0) {
            return "✔";
          }
          return count;
        },
      }),
      getColumn({
        id: "cookies",
        title: "Cookies",
        info: "Nombre de cookies présents",
        category: "best-practices",
        hash: "thirdparties",
        gradeKey: "cookiesGrade",
        sort: (a, b) => {
          if (a.summary[`cookiesCount`] === undefined) {
            return -1;
          }
          if (b.summary[`cookiesCount`] === undefined) {
            return 1;
          }
          return a.summary[`cookiesCount`] - b.summary[`cookiesCount`];
        },
        gradeLabel: (rowData) => {
          const count = rowData.summary.cookiesCount;
          if (count === 0) {
            return "✔";
          }
          return count;
        },
      }),
    ]);
  }

  if (isToolEnabled("stats")) {
    columns.push(
      getColumn({
        id: "stats",
        category: "best-practices",
        title: "Stats",
        info: "Présence de la page des statistiques",
        hash: "stats",
        gradeKey: "statsGrade",
        gradeLabel: (rowData) => {
          const grade = rowData.summary.statsGrade;
          if (grade === "A") {
            return "✔";
          }
          return grade;
        },
      })
    );
  }

  if (isToolEnabled("budget_page")) {
    columns.push(
      getColumn({
        id: "budget_page",
        category: "best-practices",
        title: "Budget",
        info: "Présence de la page de budget",
        hash: "budget_page",
        gradeKey: "budgetPageGrade",
      })
    );
  }

  if (isToolEnabled("404")) {
    columns.push(
      getColumn({
        category: "best-practices",
        id: "404",
        title: "404",
        info: "Pages introuvables",
        hash: "404",
        gradeKey: "404",
        gradeLabel: (rowData) => {
          const count = rowData.summary["404"];
          //@ts-ignore
          if (count === 0 || count === "A+") {
            return "✔";
          }
          return count;
        },
      })
    );
  }

  if (isToolEnabled("trivy")) {
    columns.push(
      getColumn({
        id: "trivy",
        category: "securite",
        title: "Trivy",
        info: "Vulnérabilités Trivy",
        hash: "trivy",
        gradeKey: "trivyGrade",
      })
    );
  }

  if (isToolEnabled("sonarcloud")) {
    columns.push(
      getColumn({
        id: "sonarcloud",
        category: "securite",
        title: "SonarCloud",
        info: "Scan de code SonarCloud.io",
        hash: "sonarcloud",
        gradeKey: "sonarcloudGrade",
      })
    );
  }

  const filterBy = (key) => (item, idx, arr) =>
    !arr.slice(idx + 1).find((r) => item[key] === r[key]);
  return (
    (report && (
      <Table
        data={report.filter(filterBy("url"))}
        caption={""}
        columns={columns}
        rowKey={(row) => row.url}
        perPage={1000}
        tableClassName={styles.table}
        className={styles.tableWrapper}
        fixedHeader={true}
      />
    )) ||
    null
  );
};
