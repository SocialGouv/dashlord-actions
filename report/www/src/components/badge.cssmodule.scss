.success,
.info,
.warning,
.danger {
    display: inline-block;
    cursor: pointer;
    padding: 4px 8px;
    color: white;
    min-width: 60px;
    border-radius: 4px;
    text-align: center;
    font-weight: bold;
}

.success {
    background-color: var(--success);
}

.info {
    background-color: var(--info);
}

.warning {
    background-color: #FF6F4C;
}

.danger {
    background-color: var(--error);
}