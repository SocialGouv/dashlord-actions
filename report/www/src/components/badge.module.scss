.success,
.info,
.warning,
.danger {
    display: inline-block;
    cursor: pointer;
    padding: 4px 8px;
    color: white;
    min-width: 60px;
    border-radius: 4px;
    text-align: center;
    font-weight: bold;
}



.success, .success:hover {
    background-color: var(--success) !important;
}

.info, .info:hover {
    background-color: var(--info) !important;
}

.warning, .warning:hover {
    background-color: var(--warning) !important;
}

.danger, .danger:hover {
    background-color: var(--error) !important;
}