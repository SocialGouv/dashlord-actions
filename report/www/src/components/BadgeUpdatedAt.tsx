import { formatDistanceToNow, differenceInDays, format } from "date-fns";
import { fr } from "@codegouvfr/react-dsfr";
import frLocale from "date-fns/locale/fr";

import { AlertProps } from "@codegouvfr/react-dsfr/Alert";
import Badge from "@codegouvfr/react-dsfr/Badge";
import { CSSProperties } from "react";

// badge that displya "since xxx" and appropriate colors
export const BadgeUpdatedAt = ({
  date,
  style,
}: {
  date: string;
  style: CSSProperties;
}) => {
  if (!date) return;
  const startDate = new Date(date);
  const lagInDays = differenceInDays(new Date(), startDate);
  let severity: AlertProps.Severity = "info";
  if (lagInDays > 60) {
    severity = "error";
  } else if (lagInDays > 30) {
    severity = "warning";
  }
  return (
    <Badge
      severity={severity}
      noIcon
      as="span"
      style={{ verticalAlign: "top", ...style }}
    >
      <i
        className={fr.cx("fr-icon-time-fill", "fr-icon--sm", "fr-mr-1v")}
        title={`Mis Ã  jour le ${format(new Date(startDate), "dd/M/yyyy")}`}
      />
      {formatDistanceToNow(new Date(startDate), {
        locale: frLocale,
      })}
    </Badge>
  );
};
