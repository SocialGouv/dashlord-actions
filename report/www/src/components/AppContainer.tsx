import * as React from "react";

import { Container, Alert } from "@dataesr/react-dsfr";

import { Switch, Route, useLocation, useParams } from "react-router-dom";

import { Dashboard } from "./Dashboard";
import { Trends } from "./Trends";
import { Intro } from "./Intro";
import { About } from "./About";
import { WappalyzerDashboard } from "./WappalyzerDashboard";
import Url from "./Url";

const report: DashLordReport = require("./report.json");
const trends: Trends = require("./trends.json");

type CategoryRouteProps = { report: DashLordReport };

// for some reason react-router `:url*` didnt work, use `*` only
interface CategoryParamTypes {
  category: string;
}

const CategoryRoute: React.FC<CategoryRouteProps> = (props) => {
  const params = useParams<CategoryParamTypes>();
  const category = window.decodeURIComponent(params.category);
  const urls = props.report.filter((u) => u.category === category);
  return (
    <>
      <br />
      {urls.length ? (
        <>
          <h3>
            {category} :{urls.length} urls
          </h3>
          <Dashboard report={urls} />
        </>
      ) : (
        <h3>Aucun URL associ√©e</h3>
      )}
    </>
  );
};

interface TagParamTypes {
  tag: string;
}

type TagRouteProps = { report: DashLordReport };

const TagRoute: React.FC<TagRouteProps> = (props) => {
  const params = useParams<TagParamTypes>();
  const tag = window.decodeURIComponent(params.tag);
  const urls = props.report.filter((u) => u.tags && u.tags.includes(tag));
  return (
    <>
      <br />
      <h3>
        {tag} :{urls.length} urls
      </h3>
      <Dashboard report={urls} />
    </>
  );
};

interface UrlParamTypes {
  "0": string;
}

type UrlRouteProps = { report: DashLordReport };

const UrlRoute: React.FC<UrlRouteProps> = (props) => {
  const params = useParams<UrlParamTypes>();
  const url = window.decodeURIComponent(params["0"]);
  const urlData = props.report.find((u) => u.url === url);
  if (!urlData) {
    return (
      <Alert
        type="error"
        title={`Impossible de trouver le rapport pour ${url}`}
      />
    );
  }
  return <Url url={url} report={urlData} />;
};

export const AppContainer = () => {
  const location = useLocation();
  return (
    <Container fluid={location.pathname === "/"}>
      <div role="main" className="fr-my-4w">
        <Switch>
          <Route path="/url/*">
            <UrlRoute report={report} />
          </Route>
          <Route path="/dashboard">
            <Dashboard report={report} />
          </Route>
          <Route path="/trends">
            <Trends trends={trends} />
          </Route>
          <Route path="/category/:category">
            <CategoryRoute report={report} />
          </Route>
          <Route path="/tag/:tag">
            <TagRoute report={report} />
          </Route>
          <Route path="/wappalyzer">
            <WappalyzerDashboard report={report} />
          </Route>
          <Route path="/intro">
            <Intro />
          </Route>
          <Route path="/about">
            <About />
          </Route>
          <Route path="/">
            <Dashboard report={report} />
          </Route>
        </Switch>
      </div>
    </Container>
  );
};
