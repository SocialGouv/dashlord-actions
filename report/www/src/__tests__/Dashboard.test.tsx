/// <reference types="vitest" />

import path from "path";
import * as React from "react";
import * as renderer from "react-test-renderer";

// jest.mock("next/router", () => jest.requireActual("next-router-mock"));

// jest.mock("@mui/x-data-grid/components/GridColumnHeaders", () => {
//   return <div>pouet</div>;
// });

//import { GridColumnHeaders } from "@mui/x-data-grid";

// const spyGridColumnHeaders = jest.spyOn(
//   require("@mui/x-data-grid"),
//   "GridColumnHeaders"
// );
// spyGridColumnHeaders.mockImplementation(() => <div>pouet</div>);

// jest.mock("@mui/x-data-grid", () => () => {
//   return <div>header</div>;
// });

import { Dashboard } from "../components/Dashboard";

import report from "../report.json";

import fs from "fs";
import { vi } from "vitest";

vi.mock("@/config.json", async (importOriginal) => {
  const config = fs
    .readFileSync(path.join(__dirname, "config.json"))
    .toString();
  console.log("config", config);
  return config;
});

vi.mock("@mui/x-data-grid", async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    DataGrid: () => <div>empty dashboard</div>, // .fn(() => null),
  };
  //return config;
});

it("Should render empty Dashboard", () => {
  const props = {};
  const tree = renderer.create(<Dashboard {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("Should render full Dashboard", () => {
  const props = { report };
  const tree = renderer.create(<Dashboard {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("Tools config", () => {
  beforeEach(() => {
    vi.mock("../../config.json", async (importOriginal) => {
      const config = fs
        .readFileSync(path.join(__dirname, "config.json"))
        .toString();
      return {
        ...config,
        tools: { http: false, zap: true, lighthouse: false },
      };
    });
  });

  it("Should render Dashboard with limited tools", () => {
    const props = { report };
    const tree = renderer.create(<Dashboard {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
