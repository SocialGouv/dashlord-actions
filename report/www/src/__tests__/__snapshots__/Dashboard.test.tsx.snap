// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render empty Dashboard 1`] = `null`;

exports[`Should render full Dashboard 1`] = `
<div>
  empty dashboard
</div>
`;

exports[`Tools config > Should render Dashboard with limited tools 1`] = `
<div>
  empty dashboard
</div>
`;
